this qualified field masked by argument

I can't say definitively this is a bug.  But I think it should be.  There is only one thing which "this.x" can mean.  It looks to me like this could be implemented with some extra logic in typedAssign.
```scala
class A(var x: Int)
class B(x: Int) extends A(x) { def update(z: Int) = this.x = z }
```
```scala
a.scala:2: error: reassignment to val
class B(x: Int) extends A(x) { def update(z: Int) = this.x = z }
                                                           ^
one error found
```
I would expect it to work like the common situation (common in java anyway) where one does:
```scala
def f(x: Int) = this.x = x
```

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -4573,16 +4573,27 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         if (varsym == null)
           return fail()
 
-        if (treeInfo.mayBeVarGetter(varsym)) {
+        def shadowsSetter =
+          lhs1 match {
+            case treeInfo.Applied(Select(qual, _), _, _) =>
+              qual.tpe.member(varsym.name.setterName).exists
+            case _ => false
+          }
+
+        def setterRewrite =
           lhs1 match {
             case treeInfo.Applied(Select(qual, _), _, _) =>
               val sel = Select(qual, varsym.name.setterName) setPos lhs.pos
               val app = Apply(sel, List(rhs)) setPos tree.pos
-              return typed(app, mode, pt)
-
-            case _ =>
+              typed(app, mode, pt)
+            case _ => EmptyTree
           }
+
+        if (treeInfo.mayBeVarGetter(varsym)) {
+          val res = setterRewrite
+          if (!res.isEmpty) return res
         }
+
 //      if (varsym.isVariable ||
 //        // setter-rewrite has been done above, so rule out methods here, but, wait a minute, why are we assigning to non-variables after erasure?!
 //        (phase.erasedTypes && varsym.isValue && !varsym.isMethod)) {
@@ -4595,6 +4606,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             Apply(lhs1, List(rhs))
           }
           wrapErrors(t, _.typed1(t, mode, pt))
+        }
+        else if (shadowsSetter) {
+          val res = setterRewrite
+          if (!res.isEmpty) res else fail()
         }
         else fail()
       }

--- a/test/files/run/t3194.scala
+++ b/test/files/run/t3194.scala
@@ -1,0 +1,11 @@
+// scalac: -language:_
+import scala.util.chaining._
+
+class A(var x: Int)
+class B(x: Int) extends A(x) {
+  def update(v: Int): Unit = x = v
+}
+
+object Test extends App {
+  new B(27).tap(_.update(42)).ensuring(_.x == 42)
+}

